java/src/main/kotlin/io/spine/validation/java/JavaValidationRenderer.kt

```diff
@@ -52,86 +51,75 @@ import io.spine.validation.java.point.ValidateBeforeReturn
-        sources.annotateGeneratedMessages(messageTypes)
-        sources.plugValidationIntoBuild(messageTypes)
-    }
 
-    private fun findValidations(): Validations {
-        val client = select(MessageValidation::class.java)
-        return Validations(client)
-    }
+        val allCompiledMessages = select(CompiledMessage::class.java).all()
+            .associateWith { sources.javaFileOf(it.type) }
 
-    private fun generateCode(type: MessageWithFile) {
-        val message = type.message
-        val javaFile = message.javaFile(type.fileHeader)
-        sources.find(javaFile)?.let {
-            @Suppress("UNCHECKED_CAST")
-            (it as SourceFile<Java>).addValidationCode(message)
+        // Adds `validate()` and `implements ValidatableMessage`.
+        allCompiledMessages.forEach { (message, file) ->
+            val validationCode = ValidationCode(renderer = this, message, file)
+            validationCode.generate()
         }
-    }
 
-    private fun SourceFile<Java>.addValidationCode(type: MessageType) {
-        val validation = validations[type]
-        val validationCode = ValidationCode(this@JavaValidationRenderer, validation, this)
-        validationCode.generate()
+        // Annotates `build()` and `buildPartial()` methods.
+        // Adds invocation of `validate()` in `build()`.
+        allCompiledMessages.values.distinct()
+            .forEach {
+                // Though, it seems logical to do this along with adding `validate()`
+                // in the `forEach` above; we cannot do that. Insertion points used
+                // for the codegen here do not contain the type name, while insertion
+                // points for the codegen above contain them.
+                //
+                // Moving these calls to `forEach` above leads to generation of
+                // the same annotations and "plugging code" several times if a compiled
+                // message has one or more nested messages (so, several `Builder` classes)
+                // within one Java file.
+                //
+                // Notice usage if `distinct()` before `forEach`. Several `MessageType`s
+                // point to the same Java file if a root message declares nested messages.
+                //
+                // We are not going to address this inconsistency here.
+                // This issue will not exist when the class migrates to PSI.
+                //
+                it.annotateBuildMethod()
+                it.annotateBuildPartialMethod()
+                it.plugValidationIntoBuild()
+            }
     }
 }
 
-/**
- * Locates source files for the given message types and
- * performs the given action for each file.
- */
-private fun SourceFileSet.forEachSourceFile(
-    messageTypes: Set<MessageWithFile>,
-    action: SourceFile<Java>.() -> Unit
-) {
-    messageTypes
-        .map { m -> m.message.javaFile(m.fileHeader) }
-        .mapNotNull { path -> find(path) }
-        .distinct()
```

Comment:
Are you sure that the newly introduced `distinct()` (line 98) is semantically equal to this whole block? I suspect there might be a subtle difference.

Reply:
Now, ProtoData combines calls of `javaFile(header)`, `find(path)` and unchecked cast to `SourceFile<Java>` in a single `SourceFileSet.javaFileOf()` method.

This new method throws when `find(path)` returns `null` (meaning no Java file). I have added ` if (!sources.hasJavaRoot) { return }` in the beginning of the renderer to avoid this.